### PR TITLE
CCurlFile::Exists() to fallback to GET if 405 is returned

### PR DESCRIPTION
**Pull request by lobermann**
## Description
I found this issue as as the iPhone App (JSON RPC) did not show any covers that where not cached by kodi. 

There can be the case that the webserver does not support the HEAD request. 
In this case (return code 405, Method not allowed)  we can fall back to a GET request to check
if the file exists. Using the range hander to only return 1 byte of the file if supported by the server.

If there can be other cases where a fallback to GET would be good, let me know.

## How Has This Been Tested?
Yes, tested on MacOS.

## Types of change
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [X] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

### CODE REVIEW

**Comment by Rechi** [comment]
jenkins build this please

**Comment by wsnipex** [comment]
this PR makes sense. Exists is called a gazillion times from everywhere and returning false because the server refuses HEAD requests is a bug.
We already have nearly the same logic(and more) in https://github.com/xbmc/xbmc/blob/4060651c5f122ac92162d054f88794a83118cd76/xbmc/filesystem/CurlFile.cpp#L1448-L1458 
so we should do the same here

**Comment by lobermann** [comment]
@peak3d I agree that magic behind the scenes can be an issue. But it can also be an issue if people do not implement the correct behaviour.
There is a byte range specified in the GET Request in order to avoid downloads of big files. 

@wsnipex thanks for the insight!

I went ahead and updated the code to include the timeout via advancedsettings and the CURLOPT_XFERINFOFUNCTION and CURLOPT_NOPROGRESS from the Stat function.

File: xbmc/filesystem/CurlFile.cpp [hunk 1]
```diff
@@ -1235,23 +1235,61 @@ bool CCurlFile::Exists(const CURL& url)
...
+      if (code == 405)
+      {
+        // If we get a Method Not Allowed response, retry with a GET Request
+        g_curlInterface.easy_setopt(m_state->m_easyHandle, CURLOPT_NOBODY, 0);
+        
+        g_curlInterface.easy_setopt(m_state->m_easyHandle, CURLOPT_FILETIME, 1);
+#if LIBCURL_VERSION_NUM >= 0x072000 // 0.7.32
+        g_curlInterface.easy_setopt(m_state->m_easyHandle, CURLOPT_XFERINFOFUNCTION, transfer_abort_callback);
+#else
+        g_curlInterface.easy_setopt(m_state->m_easyHandle, CURLOPT_PROGRESSFUNCTION, transfer_abort_callback);
```
[the remark this answers is not in the document]
Comment (lobermann) [reply xbmc/filesystem/CurlFile.cpp]: ok, thanks. removed.

[the remark this answers is not in the document]
Comment (wsnipex) [reply xbmc/filesystem/CurlFile.cpp]: I think @rechi meant the curl version specific line is not needed. the transfer_abort_callback should definitely be used though.

**Review by wsnipex — APPROVED** [review]

**Comment by wsnipex** [comment]
@peak3d ok for you?

**Review by peak3d — APPROVED** [review]

**Comment by lobermann** [comment]
Thanks @Rechi @wsnipex @peak3d . If there are no more comments I will go ahead and merge it tomorrow evening.

**Comment by peak3d** [comment]
Yes, cool, thx for contribution!

**Pull request merged**